### PR TITLE
[llvm-cov] Coverage report HTML UI to jump between uncovered parts of code

### PR DESCRIPTION
**Pull request by hanickadot**
I replaced "jump to first uncovered line" with UI buttons and keyboard shortcut to jump between uncovered parts of code: lines (key L), branchs (key B), regions (key R).

User can also jump in reverse direction with shift+key.

### CODE REVIEW

**Comment by hanickadot** [comment]
Motivation: I recently was writing tests for uncovered code in a long file and I spend a lot of time scrolling and looking for missing uncovered lines inside instantiation (to which the link "jump to first uncovered line" didn't even work.

**Comment by hanickadot** [comment]
<details>
  <summary>previous</summary>
https://youtu.be/B48yDtQq2FE

</details>
<details>
  <summary>after</summary>
  https://www.youtube.com/watch?v=LEgrRTa5B4E
</video>

</details>

**Comment by hanickadot** [comment]
Screenshot before:
Screenshot 2024-06-15 at 17 35 43
[video link](https://youtu.be/B48yDtQq2FE)

Screenshot after:
Screenshot 2024-06-15 at 17 35 06
[video link](https://youtu.be/LEgrRTa5B4E)

**Comment by hanickadot** [comment]
Demo here: https://tmp.cpp.fail/report/coverage/include/cthash/sha3/common.hpp.html

**Comment by ornata** [comment]
Cool! I tested in Firefox, Safari, and Chrome and the feature works in all 3.

**Review by ornata — APPROVED** [review]

File: llvm/tools/llvm-cov/SourceCoverageViewHTML.cpp [hunk 1]
```diff
@@ -88,13 +88,123 @@ const char *BeginHeader =
     "<meta name='viewport' content='width=device-width,initial-scale=1'>"
     "<meta charset='UTF-8'>";
 
+const char *JSForCoverage =
+    R"javascript(
+
+function next_uncovered(selector, reverse, scroll_selector) {
```
Comment (chapuni) [inline llvm/tools/llvm-cov/SourceCoverageViewHTML.cpp:94]: Could we unify the cursor? I guess users want to jump to the prev/next *visible* issue.
For example, "Jump to the prev/next another uncovered line in sight when the cursor is navigated by branch."

**Review by chapuni — CHANGES_REQUESTED** [review]

**Comment by chapuni** [comment]
Could we split out `.css` and `.js` to individual include files in the future? (I don't think we should do right now)

File: llvm/tools/llvm-cov/SourceCoverageViewHTML.cpp (hunk 1 above)
Comment (hanickadot) [reply llvm/tools/llvm-cov/SourceCoverageViewHTML.cpp:94]: I don't know how, at least not without really big refactor in how the HTML is generated.

**Comment by hanickadot** [comment]
> Could we split out `.css` and `.js` to individual include files in the future? (I don't think we should do right now)

It is, at least for me I have split .css and .js included by filepath and not as content. Or do you mean something else?

**Comment by hanickadot** [comment]
> Could we split out `.css` and `.js` to individual include files in the future? (I don't think we should do right now)

Oh, I wish we could just #embed it :)

File: llvm/tools/llvm-cov/SourceCoverageViewHTML.cpp [hunk 2]
```diff
@@ -487,6 +651,9 @@ Error CoveragePrinterHTML::createIndexFile(
   if (Error E = emitStyleSheet())
     return E;
 
+  if (Error E = emitJavaScript())
+    return E;
+
```
[the remark this answers is not in the document]
Comment (hanickadot) [reply llvm/tools/llvm-cov/SourceCoverageViewHTML.cpp:657]: thanks, fixed!

**Review by chapuni — APPROVED** [review]
Ok-maru-suisan, thanks! @hanickadot

**Pull request merged**